src/core/execute.c

```diff
@@ -2568,82 +2693,69 @@ static int acquire_credentials(
 
-                if (lc->encrypted) {
-                        _cleanup_free_ void *plaintext = NULL;
-                        size_t plaintext_size = 0;
+                if (!is_dir) {
+                        r = load_credential(context, params, lc, unit, dfd, uid, ownership_ok, &left);
+                        if (r < 0)
+                                return r;
+
+                } else {
+                        char **path;
+                        _cleanup_strv_free_ char **credpaths = NULL;
+
+                        r = recurse_dir_at(
+                                        dfd,
+                                        lc->path,
+                                        /* statx_mask= */ 0,
+                                        /* n_depth_max= */ UINT_MAX,
+                                        RECURSE_DIR_IGNORE_DOT|RECURSE_DIR_ENSURE_TYPE,
+                                        append_path_recurse_dir_cb,
+                                        &credpaths);
 
-                        r = decrypt_credential_and_warn(lc->id, now(CLOCK_REALTIME), NULL, data, size, &plaintext, &plaintext_size);
                         if (r < 0)
                                 return r;
 
-                        free_and_replace(data, plaintext);
-                        size = plaintext_size;
-                }
+                        STRV_FOREACH(path, credpaths) {
+                                _cleanup_(exec_load_credential_freep) ExecLoadCredential *sub_lc = NULL;
+                                _cleanup_free_ char *cred = NULL, *sub_id = NULL;
 
-                add = strlen(lc->id) + size;
-                if (add > left)
-                        return -E2BIG;
+                                r = path_make_relative(lc->path, *path, &cred);
+                                if (r < 0)
+                                        return r;
+                                sub_id = strreplace(cred, "/", "_");
 
-                r = write_credential(dfd, lc->id, data, size, uid, ownership_ok);
-                if (r < 0)
-                        return r;
+                                sub_lc = new(ExecLoadCredential, 1);
+                                if (!sub_lc)
+                                        return -ENOMEM;
 
-                left -= add;
+                                *sub_lc = (ExecLoadCredential) {
+                                        .id = strjoin(lc->id, "_", sub_id),
+                                        .path = TAKE_PTR(*path),
+                                        .encrypted = lc->encrypted,
+                                };
+
+                                if (!sub_lc->id)
+                                        return -ENOMEM;
+
+                                r = load_credential(context, params, sub_lc, unit, dfd, uid, ownership_ok, &left);
+                                if (r < 0)
+                                        return r;
```

Comment:
any reason you don't just call load_credential() from the recursion callback? what's the reason for first generating a list, and then processing that in a second step?

I'd feel more comfortable btw, if we'd no go by path to load the credential btw. i.e. inside the recursion callback you get the directory fd and directory entry name and then you can do openat(directory_fd, directory_entry->name, …) to open the credential. We should use that fact. It's safer against races, i.e. with other stuff changing the dir we read from.

read_full_file_full() (i.e. the main part of load_credential()) supports openat() internally too, just pass the first two params appropriately